File: 238. Product of Array Except Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> res;
        int n = nums.size();
        if (n == 0) return res;
        res.resize(n);

        int i, product = 1;
        for (i = n-1; i >= 0; i--) {
            res[i] = product;
            product *= nums[i];
        }
        
        product = 1;
        for (i = 0; i < n; i++) {
            res[i] *= product;
            product *= nums[i];
        }
        
        return res;
    }
};

File: python/238. Product of Array Except Self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        n = len(nums)
        output = [1] * n
        product = 1
        for i in range(n-1, -1, -1):
            output[i] *= product
            product *= nums[i]
        
        product = 1
        for i in range(n):
            output[i] *= product
            product *= nums[i]
        return output
            